src/nvim/normal.c

```diff
@@ -1290,6 +1290,71 @@ static void normal_check_cursor_moved(NormalState *s)
   }
 }
 
+static int last_visualchanged_mode;
+static pos_T last_visualchanged_start;
+static pos_T last_visualchanged_end;
+
+/// Trigger VisualChanged if visual selection has changed
+static void normal_check_visual_changed(void)
+{
+  if (!finish_op && VIsual_active && has_event(EVENT_VISUALCHANGED)) {
+    pos_T start = VIsual;
+    pos_T end = curwin->w_cursor;
+    if (lt(end, start)) {  // swap
+      start = curwin->w_cursor;
+      end = VIsual;
+    }
+
+    if (VIsual_mode == Ctrl_V) {
+      if (start.col > end.col) {
+        // swap needed because shift-o might change col
+        // even though selection remains the same
```

Comment:
Do we really need to swap here? Typically, selection has a direction (active and anchor).

Reply:
I don't think the direction is relevant because the range remains the same in the end.
I tried to "normalize" the range by having the start as the top-left corner and the end as bot-right.
In this case, the corners stored in `w_cursor` and `VIsual` are top-right and bot-left. Without the swap, the event will fire when the cursor moves to another corner without actually changing the range.

Reply:
I think VisualChanged should consider both range change and anchor change, otherwise this event is incomplete and the data of v:event is useless. For example, in VSCode, a selection is defined by its starting and ending positions. Additionally, within selections, there is the concept of a primary selection (In Vim's visual block mode, the selection under the current cursor position is referred to as the primary selection).

Reply:
Added anchor and cursor as byte-index, should the UI event include virtcol or other data?